lib/spack/spack/cmd/spec.py

```diff
@@ -98,13 +110,16 @@ def spec(parser, args):
             continue
 
         with tree_context():
-            kwargs['hashes'] = False  # Always False for input spec
-            print("Input spec")
-            print("--------------------------------")
-            print(spec.tree(**kwargs))
+            # Only show the Input / Concretized headers for specs that are not
+            # concrete to avoid repeated output.
+            if not spec.concrete:
+                kwargs['hashes'] = False  # Always False for input spec
+                print("Input spec")
+                print("--------------------------------")
+                print(spec.tree(**kwargs))
+                print("Concretized")
+                print("--------------------------------")
```

Comment:
Does this mean that `spack spec foo/hash` will no longer print the input spec? I'm not sure that's a change we would want.

Reply:
I think if the spec is concrete but was read in from the command line, we want the value from the command line here. But I agree about not printing the abstract spec for each concrete spec in an environment here.

Reply:
Yeah, I'm not sure what to do about that to be honest. `spack.spec.parse('/xyz')` returns a concrete spec from the database, so it's rather "find" than "parse".